### ISSUE TITLE
gbasis and mocoeffs

### ISSUE THREAD

**Issue by felixplasser** [cclib/cclib#748]
Hello, I was playing around with the gbasis and mocoeffs attributes a little bit. I was hoping to create a script  ( [cc2molden](https://sourceforge.net/p/theodore-qc/code/ci/master/tree/bin/cc2molden.py) ) that allows me to automatically convert from cclib to Molden format.

I just wanted to mention two points where I ran into trouble: First, cclib does not say whether we have spherical or Cartesian basis functions. Second, different quantum chemistry codes use a different ordering of the basis functions for any given l-value (e.g. p_x, p_y, p_z). So, generally speaking, the orbitals are not uniquely defined by gbasis and mocoeffs. Would it make sense to mention that in the "Parsed data notes"?

I think PyQuante would run into the same problem. Or is there some mechanism that assures that the ordering of the basis functions is always the same?

Cheers
Felix

**Comment by berquist** [comment]
Hi Felix,

We have a Molden writer: https://github.com/cclib/cclib/blob/master/cclib/io/moldenwriter.py

1. It looks like we assume all basis functions are Cartesian and not spherical when performing the writing, and we don't generally handle this internally via an attribute or some metadata. There are some programs that maybe do internal fiddling if even parsing the MOs is problematic:
```
$ rg -Fi 'spherical' cclib
cclib/parser/gamessparser.py
1270:        elif line.find("SPHERICAL HARMONICS KEPT IN THE VARIATION SPACE") >= 0:

cclib/parser/nwchemparser.py
201:        # It also appears that we have to handle cartesian vs. spherical
1101:        if self.shells['type'] == 'spherical':

cclib/parser/molproparser.py
33:    # grows and there are both Cartesian and spherical variants to consider.
35:    # spherical ones are called 3d0, 3d1-, 3d1+, 3d2- and 3d2+. For F orbitals, the Cartesians
36:    # are xxx, xxy, xxz, xyy, ... and the sphericals are 4f0, 4f1-, 4f+ and so on.
42:        # For spherical functions, we need to construct the names.
44:        spherical = [pre + '0'] + [pre + str(j) + s for j in range(1, i+3) for s in ['-', '+']]
45:        atomic_orbital_names[orb] = cartesian + spherical
339:                # shell type (S,P,D,F,G). Molpro names also differ between Cartesian/spherical representations.

cclib/parser/qchemparser.py
328:                        # m_l when using spherical functions.

cclib/io/wfxwriter.py
246:        Spherical basis are not currently supported by the writer.

cclib/parser/psi4parser.py
236:                # with corrections for the cartesian/spherical cases.
266:        #    spherical
304:            spherical = line.strip() == "spherical"
305:            if hasattr(self, 'spherical_basis'):
306:                assert self.spherical_basis == spherical
308:                self.spherical_basis = spherical
```

2. The ordering of $m_l$ within a shell is almost certainly something that we don't handle consistently. This means
> So, generally speaking, the orbitals are not uniquely defined by gbasis and mocoeffs

is correct. Yes, this should go into the documentation. Can you go more into detail about the PyQuante comment? I don't fully understand what you're referencing.

**Comment by felixplasser** [comment]
Ok, thanks for your response. Regarding PyQuante, I was just mentioning it because it is explicitly mentioned on the documentation:
https://cclib.github.io/data_notes.html#gbasis
I think that PyQuante will have the same problems as moldenwriter.py . If the orbitals are not uniquely defined by cclib, then PyQuante can also not be assured to work properly.

By the way: In principle the necessary information is encoded in `aonames`. So, for example we know that `C1_10D 0` would be a spherical d-function. And the `aonames` also tell us the order of the different m_l within a shell.

**Comment by berquist** [comment]
Yes, I understand now. We definitely need an internal attribute for Cartesian/spherical.

**Comment by berquist** [comment]
Looks like https://github.com/cclib/cclib/issues/333 is more relevant now.
